pkg/pods/manifest.go

```diff
@@ -28,31 +27,132 @@ type LaunchableStanza struct {
+	plaintext []byte
+	signature []byte
+}
+
+// assert unsignedManifest implements Manifest
+var _ Manifest = &unsignedManifest{}
+
+// assert signedManifest implements Manifest and SignedManifest
+var _ Manifest = &signedManifest{}
+var _ SignedManifest = &signedManifest{}
+
+type unsignedManifest struct {
 	Id                string                      `yaml:"id"` // public for yaml marshaling access. Use ID() instead.
 	RunAs             string                      `yaml:"run_as,omitempty"`
 	LaunchableStanzas map[string]LaunchableStanza `yaml:"launchables"`
 	Config            map[interface{}]interface{} `yaml:"config"`
 	StatusPort        int                         `yaml:"status_port,omitempty"`
 	StatusHTTP        bool                        `yaml:"status_http,omitempty"`
-	// these fields are required to track the original text if it was signed
-	raw       []byte
-	plaintext []byte
-	signature []byte
+
+	// Used to track the original bytes so that we don't reorder them when
+	// doing a yaml.Unmarshal and a yaml.Marshal in succession
+	raw []byte
 }
 
-func (manifest *Manifest) ID() string {
+func NewUnsignedManifest() Manifest {
+	return &unsignedManifest{}
+}
+
+func (manifest *unsignedManifest) ID() string {
 	return manifest.Id
 }
 
-func (manifest *Manifest) RunAsUser() string {
+func (manifest *unsignedManifest) SetID(id string) error {
+	manifest.Id = id
+	return nil
+}
+
+func (manifest *signedManifest) SetID(id string) error {
+	return util.Errorf("Cannot change id on a signed manifest")
+}
+
+func (manifest *unsignedManifest) GetLaunchableStanzas() map[string]LaunchableStanza {
+	return manifest.LaunchableStanzas
+}
+
+func (manifest *unsignedManifest) AddLaunchableStanza(launchableId string, stanza LaunchableStanza) {
+	if manifest.LaunchableStanzas == nil {
+		manifest.LaunchableStanzas = make(map[string]LaunchableStanza)
+	}
+
+	manifest.LaunchableStanzas[launchableId] = stanza
+}
+
+func (manifest *unsignedManifest) GetConfig() map[interface{}]interface{} {
+	return manifest.Config
+}
+
+func (manifest *unsignedManifest) SetConfig(config map[interface{}]interface{}) {
+	manifest.Config = config
+}
+
+func (manifest *unsignedManifest) GetStatusPort() int {
+	return manifest.StatusPort
+}
+
+func (manifest *unsignedManifest) SetStatusPort(port int) {
+	manifest.StatusPort = port
+}
+
+func (manifest *unsignedManifest) GetStatusHTTP() bool {
+	return manifest.StatusHTTP
+}
+
+func (manifest *unsignedManifest) SetStatusHTTP(statusHTTP bool) {
+	manifest.StatusHTTP = statusHTTP
+}
+
+func (manifest *unsignedManifest) RunAsUser() string {
 	if manifest.RunAs != "" {
 		return manifest.RunAs
 	}
 	return manifest.ID()
 }
 
-func ManifestFromPath(path string) (*Manifest, error) {
+func (manifest *unsignedManifest) SetRunAsUser(user string) error {
+	manifest.RunAs = user
+	return nil
+}
+
+func (manifest *signedManifest) SetRunAsUser(user string) error {
+	return util.Errorf("Cannot change RunAsUser for a signed manifest")
```

Comment:
If we make it so that `signedManifest` doesn't even have a `SetRunAsUser` method, we can protect against this error at compile-time, not run-time.